File: com.aware.plugin.sentiment/com.aware.plugin.sentiment/src/main/java/com/aware/plugin/sentiment/SentimentAnalysis.java
package com.aware.plugin.sentiment;

import android.content.Context;
import android.util.Log;

import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

import java.io.IOException;
import java.io.InputStream;
import java.util.HashMap;
import java.util.Iterator;
import java.util.Set;

public class SentimentAnalysis {
	//confirm that context is needed
    private Context context;
    private JSONObject Dictionary;
    //store score as hashmap of categories
    private HashMap CalcScore=new HashMap<String,Double>();   ;

    public SentimentAnalysis(Context c) {
        context = c;
        //read sentiment dictionary from assets
        try {
		    Dictionary = new JSONObject(loadJSONFromAsset(c));
            JSONArray CatArr = Dictionary.getJSONArray("categories");
            for(int i=0;i<CatArr.length();i++){
                String CatVal=CatArr.get(i).toString();
                Log.i("ABTest","In build category is "+CatVal);
                //CalcScore.put(CatVal,0);
                Double initVal=0.00;
                if (CatVal!=null){
                    CalcScore.put(CatVal,initVal);
                }
            }
            
            String totalWordsCat = "total_words";
            Double totalWordsInitVal = 0.00;
            CalcScore.put(totalWordsCat, totalWordsInitVal);
        } catch (JSONException ex) {
            ex.printStackTrace();

        }
    }
	public void resetScore(){
        //method to reset score when package is swicthed
        // Getting an iterator
        Set<String> keys = CalcScore.keySet();
        for(String key: keys){
            CalcScore.replace(key,0.0);
        }



    }
	//function to read json file to string
	public String loadJSONFromAsset(Context context) {
        String json = null;
        try {
            InputStream is;
            int id = context.getResources().getIdentifier("sentiment", "raw", context.getPackageName());
            is = context.getResources().openRawResource(id);

            int size = is.available();

            byte[] buffer = new byte[size];

            is.read(buffer);

            is.close();

            json = new String(buffer, "UTF-8");


        } catch (IOException ex) {
            ex.printStackTrace();
            return null;
        }
        return json;

    }
    /**
     * Read the dictionary
     *
     * @return Dictionary as String
     */
    public String getDictionaryAsString() {
        
        return Dictionary.toString();
    }

    public boolean isMatch(String inpWord, String targetWord){
        //functiom that returns true if target word and inpword are identical or match based on wildcard
        if (targetWord.contains("*")){
            //do regexp compare of strings
            //String escapeMask=targetWord.replace(/[\-\[\]\/\{\}\(\)\+\?\.\\\^\$\|]/g, "\\$&");
            String regexString=targetWord.replace("*", ".*");
            boolean test=inpWord.matches("(?i:"+regexString+")");
            if (test){
                Log.i("ABTest","wildcard match inp is "+inpWord +" and tgt word is "+targetWord);
            }
            return test;
        } else {
            if (inpWord.equalsIgnoreCase(targetWord)){
                Log.i("ABTest","exact match inp is "+inpWord +" and tgt word is "+targetWord);
                return true;
            } else {
                return false;
            }
        }
    }
	public HashMap getScoreFromInput(String inpWord){
        double result;
        //JSONArray CatArr=Dictionary.getJSONArray("categories");
        //process dictionary first
        String[] inpArr = inpWord.split("\\s+");
        try {
            JSONObject WordObj = Dictionary.getJSONObject("words");
            Iterator wordList=WordObj.keys();
            while(wordList.hasNext()) {
                String wordItem = wordList.next().toString();
                //System.out.println("key is "+wordItem + " ");
                //Log.i("ABTest","word in dictionary is "+wordItem);
                JSONObject WordScores= WordObj.getJSONObject(wordItem);
                Iterator catScoreList=WordScores.keys();
                while(catScoreList.hasNext()) {
                    String catScoreItem = catScoreList.next().toString();
                    double catScore=WordScores.getDouble(catScoreItem);
                    for(int j=0; j<inpArr.length;j++){
                        String inpStr=inpArr[j];
                        if (isMatch(inpStr,wordItem)){
                            double currScore=Double.parseDouble(CalcScore.get(catScoreItem).toString());
                            double newScore=currScore+catScore;
                            CalcScore.replace(catScoreItem,newScore);
                        }
                    }
                }
            }
            String totalWordsCat = "total_words";
            Double totalWordCount = new Double(inpArr.length); 
            double currTotalWordScore = Double.parseDouble(CalcScore.get(totalWordsCat).toString());
            double newTotalWordScore = currTotalWordScore + totalWordCount;
            CalcScore.replace(totalWordsCat, newTotalWordScore);

        } catch (JSONException ex){
            ex.printStackTrace();
        }

		return CalcScore;
	}

    public SentimentAnalysis getInstance(){
		return this;
	}
}

File: aware-core/src/main/java/com/aware/SentimentAnalysis.kt
package com.aware

import android.content.Context
import org.json.JSONException
import org.json.JSONObject
import java.io.IOException
import java.io.InputStream
import java.nio.charset.Charset
import edu.emory.mathcs.nlp.common.util.IOUtils
import edu.emory.mathcs.nlp.component.tokenizer.EnglishTokenizer
import edu.emory.mathcs.nlp.component.tokenizer.Tokenizer
import edu.emory.mathcs.nlp.component.tokenizer.token.Token
import org.json.JSONArray
import java.io.BufferedReader
import java.io.ByteArrayInputStream
import java.nio.charset.StandardCharsets

class SentimentAnalysis(
    context: Context,
) {

    private var context: Context
    private lateinit var dictionary: JSONObject
    private lateinit var dictionaryMap: HashMap<String, HashMap<String, Double>>
    private lateinit var categoriesArray: JSONArray
    private lateinit var sentimentDataMap: HashMap<String, Pair<Double, Int>>


    init{
        this.context = context
        try{
            dictionary = JSONObject(loadDictionary(this.context))
        }catch(e: JSONException){
            e.printStackTrace()
        }
        buildMaps()
    }

    private fun buildMaps() {

        dictionaryMap = HashMap()
        sentimentDataMap = HashMap()
        val words: JSONObject = dictionary.getJSONObject("words")
        val wordList: Iterator<String> = words.keys()
        while(wordList.hasNext()){
            val word: String = wordList.next()
            val categoryAndScore: HashMap<String, Double> = HashMap()
            val categories: JSONObject = words.getJSONObject(word)
            val categoriesList: Iterator<String> = categories.keys()
            while(categoriesList.hasNext()){
                val category: String = categoriesList.next()
                val score: Double = categories.getDouble(category)
                categoryAndScore[category] = score
            }
            dictionaryMap[word] = categoryAndScore
        }

        categoriesArray = dictionary.getJSONArray("categories")
        resetScore()
    }

    fun getScores(tokens: List<Token>){
        for(token in tokens){
            if(dictionaryMap.containsKey(token.toString())){
                val tempMap = dictionaryMap[token.toString()]
                tempMap?.map{ (category, score) ->
                    val currentScore: Double = sentimentDataMap[category]!!.first
                    val newScore = currentScore.plus(score)
                    val newCount = sentimentDataMap[category]!!.second.plus(1)
                    sentimentDataMap.replace(category, Pair(newScore, newCount))
                }
            }
        }
    }

    fun resetScore() {
        val size = categoriesArray.length()
        val initialScore = 0.0
        val initialCount = 0
        for(i in 0 until size){
            val categoryItem: String = categoriesArray[i].toString()
            sentimentDataMap[categoryItem] = Pair(initialScore, initialCount)
        }
    }

    private fun loadDictionary(context: Context): String{
        val json: String
        try{
            val inputStream: InputStream
            val identifier = context.resources.getIdentifier("sentiment_dictionary", "raw", context.packageName)
            inputStream = context.resources.openRawResource(identifier)
            val size = inputStream.available()
            val buffer = ByteArray(size)
            inputStream.read(buffer)
            inputStream.close()
            val charset: Charset = Charsets.UTF_8
            json = String(buffer, charset)
        }catch (ex: IOException){
            ex.printStackTrace()
            return "" //or null (change return type to String?)
        }
        return json
    }

    fun tokenizer(text: String): MutableList<Token>{
        val tokenizer: Tokenizer = EnglishTokenizer()
        val stream: InputStream = ByteArrayInputStream(
            text.lowercase().toByteArray(StandardCharsets.UTF_8)
        )
        val input: BufferedReader = IOUtils.createBufferedReader(stream)

        var tokens: MutableList<Token> = mutableListOf()
        var line: String?

        while (input.readLine().also{ line = it } != null) {
            tokens = tokenizer.tokenize(line)
        }

        input.close()
        return tokens
    }

    fun getInstance(): SentimentAnalysis {
        return this
    }

    fun getSentimentMap(): HashMap<String, Pair<Double, Int>>{
        return sentimentDataMap
    }
    }

